### PR TITLE
fix(muon): split the fully-gathered QKV tensor before orthogonalizing under GTP

### LINKED ISSUE #6390: [feat] GTP+Muon
GTP+Muon related stuffs, including perf optimization, bug-fix, etc..

### PR DESCRIPTION
**Pull request by fanshiqing**
- [x] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do?

**Muon runs a different update rule on `linear_qkv` depending on whether GTP is
enabled, silently.** This makes the decision layout-invariant.
---

## Problem

[The split-QKV decision tests the **local shard** row count](https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/core/optimizer/__init__.py#L803-L812). Under GTP the weight is
row-sharded, so the same weight takes different paths:

| layout | `param.shape[0]` | `% 6528` | path |
|---|---|---|---|
| TP1 | 6528 (whole) | 0 ✓ | split → **3 Newton–Schulz** |
| GTP64 | **102** (shard) | ✗ | **1 whole-matrix NS** |

Three facts make it matter:

- **The rules are far apart** — 26 % different overall, **95 % on k and v** (they are
  2.9 % of the rows, so joint orthogonalization lets `q` dominate them).
- **It does not wash out** — NS amplifies the ~1e-7 fp32 reduction-order difference
  between layouts by **~10⁴x**, and that floor is flat across condition number 1e1–1e6.
- **It is invisible** — the skip is a `logging.DEBUG` record that production never sees.

**Measured at 256 GPUs** (identical init on both arms), GTP64 vs matched TP1:
training loss **+0.173 %** — a near-constant 0.09–0.12 B token lag, **4.3x outside**
the run-to-run noise floor. GTP64 + **Adam** on the identical topology shows none, as
expected: Adam is elementwise, amplification ~1x.

## Fix

1. **Decide from the global shape** — `qkv_global_rows()`:
   `param.shape[0] * gtp_size - pad_length`.
2. **Split after the all-gather** — at 6528 rows over 64 shards the q/k/v boundaries
   fall mid-shard, so no rank can split correctly before the gather.
3. take the GTP degree from the optimizer's own group, gated on
    `is_gtp_weight_remat` -- the same signal the split path uses

## Result

image

| | GTP64 vs its own TP1 baseline | vs noise floor (−0.038 … +0.040 %) |
|---|---|---|
| before | **+0.1731 %** | **4.3x outside** |
| **after** | **−0.0342 %** | ✅ **inside** |

*(4 h from scratch, matched window 4.0–16.8 B, n=31. The residual is negative — GTP
slightly ahead — i.e. an overshoot past zero, not a leftover deficit.)*

- **Bitwise**: TP1 and GTP now produce identical qkv updates (max |diff| 0.000e+00).
- **Throughput**: 2,527 → 2,536 ms/it at GTP64 (+0.36 %, inside cross-job noise). The
  split partitions the same NS FLOPs rather than adding any, on 8 attention layers.
- **Control**: `--muon-no-split-qkv` on *both* arms also lands inside the floor
  (+0.008 %), confirming the asymmetry was the mechanism. Not the right fix — turning
  the split off costs **+0.216 %** lm loss at TP1, so GTP should gain it, not TP1 lose it.

<details>
<summary>Supporting measurements</summary>

Split vs whole-matrix NS, production settings (16 steps, `polar_express`, `spectral`,
`extra_scale_factor` 0.2), same 6528×2688 gradient:

| block | rows | relative difference |
|---|---|---|
| q | 6144 | 0.116 |
| **k** / **v** | 192 | **0.947** |
| whole | 6528 | 0.255 |

NS amplification of a 1e-7 relative input perturbation:

| cond | 1e1 | 1e3 | 1e6 |
|---|---|---|---|
| qkv 6528×2688 | 10,525x | 9,720x | 20,374x |

Also eliminated with data before landing here: expert routing, grad-norm inflation,
data assignment, loss/MTP reduction scope, router BF16 quantization, vocab-sized
weights, initialization.

</details>

---

## Pre-checks

- [x] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [x] I have added proper typing to my code
- [x] I have added relevant documentation
- [x] I have run `tools/autoformat.sh` on my PR

`tests/unit_tests/generalized_tensor_parallel/test_gtp_muon_qkv.py` — 14 cases, all
passing at world 2 and 4. Nothing existing could have caught this:
`test_muon_optimizer_qkv_split` sets `param.is_qkv` by hand on one rank, and
`test_gtp_muon.py`'s parity cases carry no qkv weight.

- `test_local_shard_would_have_disabled` — the regression test proper; pins the shape
  that used to fail, `102 × 64 = 6528`.
- `test_split_after_gather_matches_tp1` — GTP result == TP1 result, per shard.
- `test_gtp_split_differs_from_whole_matrix` — negative control, so a regression to
  the old behaviour cannot pass trivially.
- `test_row_sharded_modes_raise`, `test_padding_is_excluded`,
  `test_decision_is_layout_invariant`, `test_non_qkv_shape_is_rejected`.

`test_gtp_muon.py` passes identically before and after (6 passed / 8 skipped both
sides; the 8 need an 8-GPU host).

## Reviewer notes

- **Behaviour change**: `blockwise`/`distributed` with `muon_split_qkv=True` now
  raises. Those modes keep the weight row-sharded so no rank holds a q/k/v boundary —
  they were silently getting a different update rule. `--muon-no-split-qkv` is the
  opt-out.
- `qkv_global_rows` reads `is_qkv`/`qkv_split_shapes` (tensor-parallel attrs) **and**
  `group`/`pad_length` (GTP attrs) off the same param. Both are copied at the same
  sites today; a path copying only one set would compute `global_rows` wrongly.
- **Same pattern, not fixed here**: `mixer.in_proj` [12352, 2688] is also fused
  (`[z,x,B,C,dt]`) *and* GTP-sharded, escaping only via `use_muon=False`.
- MLA still unhandled (pre-existing `TODO(deyuf)`).

### CODE REVIEW

File: tests/unit_tests/generalized_tensor_parallel/test_gtp_muon_qkv.py [hunk 1]
```diff
@@ -0,0 +1,365 @@
...
+
+    @pytest.mark.parametrize("gtp_size", [1, 2, 4, 64])
+    def test_decision_is_layout_invariant(self, gtp_size):
+        """Same weight, every GTP degree: always splittable, always the same row count."""
+        local_rows = _GROUP // gtp_size if gtp_size <= _GROUP else 1
+        global_rows = local_rows * gtp_size
+        rows, size, splittable = _rows(_param(local_rows), _SPLIT, gtp_size)
+        assert size == gtp_size
+        assert rows == global_rows
+        assert splittable == (global_rows % _GROUP == 0)
```
[the remark this answers is not in the document]
Comment (fanshiqing) [reply tests/unit_tests/generalized_tensor_parallel/test_gtp_muon_qkv.py]: Done

File: megatron/core/optimizer/emerging_optimizers.py [hunk 2]
```diff
@@ -427,14 +437,63 @@ def scaled_orthogonalize_fn_with_gtp_remat(self, p, grad, tp_group, partition_di
...
+                    f"tp_mode='{mode}'"
+                    + (f" (auto-resolved from '{self.tp_mode}')" if self.tp_mode == "auto" else "")
+                    + "; blockwise/distributed keep the rows sharded, so the q/k/v "
+                    "boundaries are not available on any rank. These weights get "
+                    "whole-matrix Newton-Schulz, which is NOT the update rule the same "
+                    "model gets at TP1. Pass --muon-tp-mode duplicated for a "
+                    "layout-invariant qkv update, or --muon-no-split-qkv to make the "
+                    "whole-matrix rule explicit at every layout.",
+                )
+            qkv_split_shapes = None
```
[the remark this answers is not in the document]
Comment (fanshiqing) [reply megatron/core/optimizer/emerging_optimizers.py:502]: Done

File: megatron/core/optimizer/emerging_optimizers.py [hunk 3]
```diff
@@ -427,14 +437,65 @@ def scaled_orthogonalize_fn_with_gtp_remat(self, p, grad, tp_group, partition_di
...
+            ]
+            return torch.cat(qkv_grads, dim=1).view(whole_grad.shape)
+
+        if not gtp_active:
+            return orthogonalize_whole_matrix(grad)
 
         gtp_rank = get_pg_rank(gtp_remat_group)
         pad_length = getattr(p, 'pad_length', 0)
 
+        if qkv_split_shapes is not None and mode != "duplicated":
```
Comment (kwangjunahn) [inline megatron/core/optimizer/emerging_optimizers.py:482]: When GTP is active and  `_resolve_tp_mode(...)` selects `distributed` this branch clears `qkv_split_shapes`. So, I think it's better to constrain `auto` to `duplicated` for split-QKV for now? (And I see that Claude had the same comment above!)

Comment (fanshiqing) [reply megatron/core/optimizer/emerging_optimizers.py:482]: makes sense, already updated~

**Review by kwangjunahn — APPROVED** [review]

**Comment by kwangjunahn** [comment]
I just noticed while reviewing [#6326](https://github.com/NVIDIA/Megatron-LM/pull/6326) that it appears to address the same GTP QKV-splitting issue as this PR. In fact, !6326 seems to consider a broader solution covering TP/GTP shards with head-wise splits. Could you check?
@mkhona-nvidia

**Review by guihong-nv — APPROVED** [review]

File: megatron/core/optimizer/emerging_optimizers.py [hunk 4]
```diff
@@ -420,21 +430,77 @@ def scaled_orthogonalize_fn_with_gtp_remat(self, p, grad, tp_group, partition_di
...
+                ).view(num_query_groups, -1, cols)
+                for g in qkv_grads
+            ]
+            return torch.cat(qkv_grads, dim=1).view(whole_grad.shape)
 
         if not gtp_active:
-            return self.scaled_orthogonalize_fn(
-                grad, tp_group, partition_dim, tp_mode_this_group=mode
-            )
+            return orthogonalize_whole_matrix(grad)
```
Comment (deepakn94) [inline megatron/core/optimizer/emerging_optimizers.py]: I don't like this function name.

Comment (deepakn94) [reply megatron/core/optimizer/emerging_optimizers.py]: Why "whole"?

Comment (deepakn94) [reply megatron/core/optimizer/emerging_optimizers.py]: `scaled_orthogonalize_fn_with_qkv_split`?

Comment (deepakn94) [reply megatron/core/optimizer/emerging_optimizers.py]: Can we re-name the argument from `whole_grad` to `grad` too?

Comment (fanshiqing) [reply megatron/core/optimizer/emerging_optimizers.py]: `whole` meant the `input` is never a GTP_remat sharded grad.

Comment (fanshiqing) [reply megatron/core/optimizer/emerging_optimizers.py]: Will rename into the name you suggested which looks more accurate for what it does

**Review by Phlip79 — APPROVED** [review]

**Review by deepakn94 — APPROVED** [review]

**Review by FDecaYed — APPROVED** [review]
